velox/functions/prestosql/Repeat.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+class RepeatFunction : public exec::VectorFunction {
+ public:
+  static constexpr int32_t kMaxResultEntries = 10'000;
+
+  bool isDefaultNullBehavior() const override {
+    // repeat(null, n) returns an array of n nulls.
+    return false;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    const auto numRows = rows.size();
+    int32_t constantCount = 0;
+    int32_t totalCount = 0;
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    DecodedVector* countDecoded = nullptr;
+
+    auto pool = context.pool();
+    BufferPtr nulls;
+    uint64_t* rawNulls = nullptr;
+    if (args[1]->isConstantEncoding()) {
+      if (args[1]->as<ConstantVector<int32_t>>()->isNullAt(0)) {
+        // If count is a null constant, the result should be all nulls.
+        context.moveOrCopyResult(
+            BaseVector::createNullConstant(outputType, numRows, pool),
+            rows,
+            result);
+        return;
+      }
+      constantCount = args[1]->as<ConstantVector<int32_t>>()->valueAt(0);
+      try {
+        checkCount(constantCount);
+      } catch (const std::exception& e) {
+        context.setErrors(rows, std::current_exception());
+        return;
+      }
+      totalCount = constantCount * numRows;
+    } else {
+      countDecoded = decodedArgs.at(1);
+      context.applyToSelectedNoThrow(rows, [&](auto row) {
+        auto count = countDecoded->isNullAt(row)
+            ? 0
+            : countDecoded->valueAt<int32_t>(row);
+        checkCount(count);
+        totalCount += count;
+      });
+      // Allocate new vector for nulls if necessary.
+      if (countDecoded->mayHaveNulls()) {
+        nulls = allocateNulls(numRows, pool);
+        rawNulls = nulls->asMutable<uint64_t>();
+      }
+    }
+
+    // Allocate new vectors for indices, lengths and offsets.
+    BufferPtr indices = allocateIndices(totalCount, pool);
```

Comment:
@zacw7 I assume you refer to this block of allocations that will be repeated. I think this is fine.

Reply:
Got it. Thanks!

Reply:
In velox/functions/prestosql/MapZipWith.cpp I used a separate struct to hold similar sets of variable. This is one way to reduce copy-paste, but I'm not sure if it is necessary.